gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/FlowSpec.java

```diff
@@ -75,13 +78,22 @@ public class FlowSpec implements Configurable, Spec {
   /** Human-readable description of the flow spec */
   final String description;
 
-  /** Flow config as a typesafe config object */
-  final Config config;
+  /* Note that since getConfig() and getConfigAsProperties() are independent accessors, `volatile` doesn't ensure a
+  * consistent view between them. In a multi-threaded scenario one should use the following access mechanism:
+  *    FlowSpec fs = ...
+  *    synchronized (fs) {
```

Comment:
suggesting users to syncronized over the entire `FlowSpec` does not look good. 
Also, it looks a unnecessary usage of calling `getConfig` and `getConfigAsProperties` and compare both.
Both the fields can *momentarily* have different value, but that's not a problem as long as only `syncronized addProperty` method has write access to them.

Reply:
I do agree it would be overkill in the general case, wherein the user is only getting one or the other of the two.  there's no need for a single accessor to synchronize on the instance, given `volatile`.

what I believe urmi's suggesting is to lock the object if one wants to access both.  this ensures both sides agree.  perhaps such clarification could be added to the comment:
```
In a multi-threaded scenario one should use the following access mechanism:
```
since "MT scenario" applies to all our code everywhere

Reply:
in fact, the new potential to mutate via `addProperty` means that `getProperties()` should make a defensive copy to be absolutely correct.

as a reflection, there really is a lot to worry about once classes become mutable.  although I see why it may be justified, this is the reason I personally strive so hard to build immutable and state-less abstractions.

Reply:
I updated the comment. I can add the combo method but it won't require access only through the combo method so I'd have to leave the comment in anyway.